File: [0121][Easy][Best_Time_to_Buy_and_Sell_Stock]/Best_Time_to_Buy_and_Sell_Stock_2.java
// solution 2, calculation based on the price min valley and comeing with max peak.
// leetcode time     cost : 1 ms
// leetcode memory   cost : 34.7 MB 
// Time  Complexity: O(N)
// Space Complexity: O(1)
public class Solution {
    public int maxProfit(int prices[]) {
        int minprice = Integer.MAX_VALUE;
        int maxprofit = 0;
        for (int i = 0; i < prices.length; i++) {
            if (prices[i] < minprice)
                minprice = prices[i];
            else if (prices[i] - minprice > maxprofit)
                maxprofit = prices[i] - minprice;
        }
        return maxprofit;
    }
}

File: [0121][Easy][Best_Time_to_Buy_and_Sell_Stock]/Best_Time_to_Buy_and_Sell_Stock_2.py
# solution 2, one time traversal of the price list.
# leetcode time     cost : 48 ms
# leetcode memory   cost : 12.5 MB 
# Time  Complexity: O(N)
# Space Complexity: O(1)
class Solution(object):
    def maxProfit(self, prices):
        """
        :type prices: List[int]
        :rtype: int
        """
        buy, ans = float('inf'), 0
        for p in prices:
            buy, ans = min(buy, p), max(ans, p-buy)
        return ans